doc/internal/man7/deprecation.pod

```diff
@@ -0,0 +1,219 @@
+ OSSL_DEPRECATEDIN_1_1_0
+ OSSL_DEPRECATEDIN_1_1_0_FOR(msg)
+
+ OPENSSL_NO_DEPRECATED_1_0_2
+ OSSL_DEPRECATEDIN_1_0_2
+ OSSL_DEPRECATEDIN_1_0_2_FOR(msg)
+
+ OPENSSL_NO_DEPRECATED_1_0_1
+ OSSL_DEPRECATEDIN_1_0_1
+ OSSL_DEPRECATEDIN_1_0_1_FOR(msg)
+
+ OPENSSL_NO_DEPRECATED_1_0_0
+ OSSL_DEPRECATEDIN_1_0_0
+ OSSL_DEPRECATEDIN_1_0_0_FOR(msg)
+
+ OPENSSL_NO_DEPRECATED_0_9_8
+ OSSL_DEPRECATEDIN_0_9_8
+ OSSL_DEPRECATEDIN_0_9_8_FOR(msg)
+
+=head1 DESCRIPTION
+
+Deprecation of a symbol is adding an attribute to the declaration of that
+symbol (function, type, variable, but we currently only do that for
+functions in our F<< <openssl/*.h> >> header files).
+
+Removal of a symbol is not the same thing as deprecation, as it actually
+removes the symbol from public view.
+
+OpenSSL configuration supports deprecation as well as simulating removal of
+symbols from public view, and also supports doing this in terms of a
+specified OpenSSL version.
+
+Deprecation is done using attribute macros having names starting with
+B<OSSL_DEPRECATED>, used with any declaration it applies to.
+
+Simulating removal is done with guard macros having names starting with
+L<OPENSSL_NO_DEPRECATED(7)>.
+
+The implementation of a deprecated symbol is kept for two reasons:
+
+=over 4
+
+=item Kept as legacy for the deprecation period
+
+It's implemented only to have it available as long as the symbol isn't
+removed entirely (be it by explicitly removing it when it's judged that it
+has been deprecated long wnough, or because of the removal is simulated).
+These need to be guarded appropriately, as shown in the L</Implementations
+kept as legacy>.
+
+=item Kept for internal purposes
+
+The implementation doesn't need to change or be guarded, just ensure that
+the declaration is available unconditionally.  That requires the inclusion
+of an internal header file very early in the translation unit until the time
+that the symbol is explicitly removed from public view and inserted into an
+internal header file, as show in L</Implementations kept for internal
+purposes>.
+
+=back
+
+=head2 General macros
+
+I<Note: none of these macros should be used directly, please use the version
+specific macros instead>
+
+OSSL_DEPRECATED() implements the deprecated attribute if the compiler
+supports it, otherwise it's defined to nothing.  It takes one argument
+I<since> that should be set to the OpenSSL version where the attribute is
+added, and will be displayed with the deprecation warning message, for
+compilers that support user specified deprecation messages.
+
+OSSL_DEPRECATED_FOR() does the same as OSSL_DEPRECATED(), but also takes a
+second argument I<msg>, which is an additional text messages to be displayed
+with the deprecation warning along with the OpenSSL version number, for
+compilers that support user specified deprecation messages.
+
+B<OPENSSL_NO_DEPRECATED> is a macro that's generated by OpenSSL
+configuration in response to the C<no-deprecated> configuration option.
+
```

Comment:
May add: _This macro suppresses the definition of deprecated symbols._

Reply:
Done